scripts/performance/measure-performance.js

```diff
@@ -0,0 +1,195 @@
+ ********************************************************************************/
+// @ts-check
+const puppeteer = require('puppeteer');
+const fs = require('fs');
+const resolve = require('path').resolve;
+const path = resolve('./workspace');
+const profilesPath = './profiles/';
+
+const lcp = 'Largest Contentful Paint (LCP)';
+const performanceTag = braceText('Performance');
+
+let name = 'Measurement';
+let url = 'http://localhost:3000/#' + path;
+let folder = 'profile';
+let headless = true;
+let runs = 10;
+
+(async () => {
+    let defaultUrl = true;
+    const args = process.argv.slice(2);
+
+    for (let index = 0; index < args.length; index++) {
+        if (args[index].startsWith('--')) {
+            if (args[index] === '--headful') {
+                headless = false;
+                continue;
+            }
+            let next = args[index + 1];
+            if (next !== undefined && !next.startsWith('--')) {
+                if (args[index] === '--name') {
+                    name = next;
+                    index++;
+                    continue;
+                }
+                if (args[index] === '--url') {
+                    url = next;
+                    defaultUrl = false;
+                    index++;
+                    continue;
+                }
+                if (args[index] === '--folder') {
+                    folder = next;
+                    index++;
+                    continue;
+                }
+                if (args[index] === '--runs') {
+                    runs = parseInt(next);
+                    index++;
+                    continue;
+                }
+            }
+        }
+    }
+
+    if (defaultUrl) getOrCreateFolder(path);
+    getOrCreateFolder(profilesPath);
+    const folderPath = profilesPath + folder;
+    getOrCreateFolder(folderPath);
+    // Wait until the browser page is available
+    const browser = await puppeteer.launch({ headless: true });
+    const page = await browser.newPage();
+    await retryConnection(() => page.goto(url), 1000);
+    browser.close();
+
+    await measurePerformance(name, url, folderPath, headless, runs);
+})();
+
+async function measurePerformance(name, url, folder, headless, runs) {
+    const multipleRuns = runs > 1 ? true : false;
+    let runNr = 1;
+    let durations = [];
+    while (runs > 0) {
+        const browser = await puppeteer.launch({ headless: headless });
+        const page = await browser.newPage();
+
+        const file = folder + '/' + runNr + '.json';
+        await page.tracing.start({ path: file, screenshots: true });
+        await page.goto(url);
+        // This selector is for the problems indicator in the status bar, as this is one of the last elements to be rendered.
+        await page.waitForSelector('.theia-ApplicationShell', { visible: true });
+        // Prevent tracing from stopping to soon and skipping a LCP candidate
+        await delay(1000);
```

Comment:
This needs explanation: is there no true stop condition we can wait for?

Reply:
Since we need to programatically decide when to watch the tracing, the difficulty is to choose an event, where the last LCP Candidate was already recorded. This is close to the time, where the `-theia-ApplicationShell` selector is visible. However, due to the loading indicator fading away it takes a little bit extra time for the tracing to realize that there is nothing more being painted. Hence, we need to wait a little bit longer. Originally i thought about waiting for a selector which comes later in the DOM, but we cannot be really sure, if this selector exists for other custom theia applications, which should also be able to use the script.